cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/KafkaNotificationService.java

```diff
@@ -0,0 +1,345 @@
+              }
+            }
+          }
+        } finally {
+          topicsFeedsLock.writeLock().unlock();
+        }
+      }
+    };
+  }
+
+  /**
+   * Callback class called when a Kafka message is received. The {@link #onReceived} method will
+   * extract the feed ID of the message received, and pass the notification encoded in the message
+   * to all handlers that are interested in that feed, using the {@code feedsToHandlers} map.
+   * One callback is created per TopicPartition. It is created by the first subscription to a feed
+   * which maps to the TopicPartition.
+   */
+  private final class KafkaNotificationsCallback implements KafkaConsumer.MessageCallback {
+
+    private final TopicPartition topicPartition;
+    private final NotificationFeed feed;
+    private final NotificationHandler handler;
+
+    private KafkaNotificationsCallback(TopicPartition topicPartition, NotificationFeed feed,
+                                       NotificationHandler handler) {
+      this.topicPartition = topicPartition;
+      this.feed = feed;
+      this.handler = handler;
+    }
+
+    @Override
+    public void onReceived(Iterator<FetchedMessage> messages) {
+      int count = 0;
+      while (messages.hasNext()) {
+        FetchedMessage message = messages.next();
+        ByteBuffer payload = message.getPayload();
+
+        topicsFeedsLock.readLock().lock();
+        try {
+          KafkaMessage decodedMessage = KafkaMessageCodec.decode(payload);
+          for (NotificationFeed feedForTopic : topicsToFeeds.get(topicPartition)) {
+            // Find the NotificationFeed that the received notification - encoded in the message - belongs to
+            if (!decodedMessage.getMessageKey().equals(
+              KafkaNotificationUtils.buildKafkaMessageKey(feedForTopic))) {
+              continue;
+            }
+
+            for (NotificationHandler handlerForFeed : feedsToHandlers.get(feedForTopic)) {
+              try {
+                // All handlers of one feed must handle notifications of the same type, otherwise some handlers
+                // will fail to decode the notification
+                Object notification = GSON.fromJson(decodedMessage.getNotificationJson(),
+                                                    handlerForFeed.getNotificationFeedType());
+                try {
+                  handlerForFeed.received(notification,
+                                          new BasicNotificationContext(dsFramework, transactionSystemClient));
+                  count++;
+                } catch (Throwable t) {
+                  LOG.warn("Error while processing notification {} with handler {}",
+                           notification, handlerForFeed, t);
+                }
+              } catch (JsonSyntaxException e) {
+                LOG.info("Could not decode Kafka message '{}' using Gson for handler {}. " +
+                           "Make sure that the getNotificationFeedType() method is correctly set.",
```

Comment:
using "()" and "method" is redundant. 
"Make sure that the getNotificationFeedType method is correctly set."